File: app/src/main/java/com/leo/wan/base/BaseActivity.java
package com.leo.wan.base;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

/**
 * @Description:
 * @Author: ch
 * @CreateDate: 2019/7/30 20:37
 */
public abstract class BaseActivity<V extends ViewDataBinding, VM extends BaseViewModel> extends AppCompatActivity {

    protected V mBinding;
    protected VM mViewModel;
    protected Context mContext;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initViewDataBinding();
        mContext = this.getApplicationContext();
        initData();
        initViewObservable();
    }


    /**
     * 注入绑定DataBinding
     */
    private void initViewDataBinding() {
        mBinding = DataBindingUtil.setContentView(this, initContentView());
        mBinding.setVariable(initVariableId(), mViewModel = initViewModel());
        mBinding.executePendingBindings();
        mBinding.setLifecycleOwner(this);
    }

    /**
     * 初始化布局文件
     */
    public abstract int initContentView();

    /**
     * 初始化ViewModel
     */
    public abstract VM initViewModel();

    /**
     * 初始化ViewModel的id
     *
     * @return BR的id(根据对应xml文件 < data > < variable > name < / variable > < / data > 自动生成)
     */
    public abstract int initVariableId();

    /**
     * 处理LiveData的订阅方法
     */
    public abstract void initViewObservable();

    /**
     刷新布局
     */
    public void refreshLayout() {
        if (mViewModel != null) {
            mBinding.setVariable(initVariableId(), mViewModel);
            mBinding.executePendingBindings();
        }
    }

    /**
     * 初始化方法
     */
    public abstract void initData();

}

File: app/src/main/java/com/leo/wan/activity/BaseActivity.kt
package com.leo.wan.activity

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.LocaleList
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.leo.wan.util.LanguageUtil

open class BaseActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun attachBaseContext(newBase: Context) {
        val context = languageWork(newBase)
        super.attachBaseContext(context)

    }

    private fun languageWork(context: Context): Context {
        // 8.0及以上使用createConfigurationContext设置configuration
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            updateResources(context)
        } else {
            context
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private fun updateResources(context: Context): Context {
        val resources = context.resources
        val locale = LanguageUtil.getLocale(context) ?: return context
        val configuration = resources.configuration
        configuration.setLocale(locale)
        configuration.setLocales(LocaleList(locale))
        return context.createConfigurationContext(configuration)
    }


}
